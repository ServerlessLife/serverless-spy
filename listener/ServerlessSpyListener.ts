import { PrettifyForDisplay } from './PrettifyForDisplay';
import {
  DynamoDBSpyHandler,
  EventBridgeSpyHandler,
  EventBridgeRuleSpyHandler,
  S3SpyHandler,
  SnsSubscriptionSpyHandler,
  SnsTopicSpyHandler,
  SqsSpyHandler,
  FunctionRequestSpyHandler,
  FunctionConsoleSpyHandler,
  FunctionResponseSpyHandler,
  FunctionErrorSpyHandler,
} from './SpyHandlers.ts';
import { WaitForParams } from './WaitForParams';
import { DynamoDBSpyEvent } from '../common/spyEvents/DynamoDBSpyEvent';
import { EventBridgeRuleSpyEvent } from '../common/spyEvents/EventBridgeRuleSpyEvent';
import { EventBridgeSpyEvent } from '../common/spyEvents/EventBridgeSpyEvent';
import { FunctionConsoleSpyEvent } from '../common/spyEvents/FunctionConsoleSpyEvent';
import { FunctionErrorSpyEvent } from '../common/spyEvents/FunctionErrorSpyEvent';
import { FunctionRequestSpyEvent } from '../common/spyEvents/FunctionRequestSpyEvent';
import { FunctionResponseSpyEvent } from '../common/spyEvents/FunctionResponseSpyEvent';
import { S3SpyEvent } from '../common/spyEvents/S3SpyEvent';
import { SnsSubscriptionSpyEvent } from '../common/spyEvents/SnsSubscriptionSpyEvent';
import { SnsTopicSpyEvent } from '../common/spyEvents/SnsTopicSpyEvent';
import { SqsSpyEvent } from '../common/spyEvents/SqsSpyEvent';

export type ServerlessSpyListener<TSpyEvents> = {
  [P in keyof FilterConditionally<TSpyEvents, `DynamoDB#${any}`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<DynamoDBSpyEvent>>
    >
  ) => Promise<DynamoDBSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `EventBridge#${any}`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<EventBridgeSpyEvent<T>>>
    >
  ) => Promise<EventBridgeSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `EventBridgeRule#${any}`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<EventBridgeRuleSpyEvent<T>>>
    >
  ) => Promise<EventBridgeRuleSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `S3#${any}`> &
    string as `waitFor${P}`]: (
    param?: PrettifyForDisplay<WaitForParams<PrettifyForDisplay<S3SpyEvent>>>
  ) => Promise<S3SpyHandler>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `SnsSubscription#${any}`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<SnsSubscriptionSpyEvent<T>>>
    >
  ) => Promise<SnsSubscriptionSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `SnsTopic#${any}`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<SnsTopicSpyEvent<T>>>
    >
  ) => Promise<SnsTopicSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `Sqs#${any}`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<SqsSpyEvent<T>>>
    >
  ) => Promise<SqsSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `Function#${any}#Request`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<FunctionRequestSpyEvent<T>>>
    >
  ) => Promise<FunctionRequestSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `Function#${any}#Console`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<FunctionConsoleSpyEvent<T>>>
    >
  ) => Promise<FunctionConsoleSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `Function#${any}#Error`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<FunctionErrorSpyEvent<T>>>
    >
  ) => Promise<FunctionErrorSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `Function#${any}#Response`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<FunctionResponseSpyEvent<T>>>
    > //      filter?: (event: FunctionResponseSpyEvent<T>) => boolean
  ) => Promise<FunctionResponseSpyHandler<T>>;
} & {
  stop: () => void;
};

type FilterConditionally<Source, Condition> = Pick<
  Source,
  {
    [K in keyof Source]: Source[K] extends Condition ? K : never;
  }[keyof Source]
>;
