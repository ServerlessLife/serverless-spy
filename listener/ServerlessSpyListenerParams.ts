import { Credentials } from '@aws-sdk/types';

export type ServerlessSpyListenerParams = {
  credentials?: Credentials;
  debugMode?: boolean;
  scope: string;
};
