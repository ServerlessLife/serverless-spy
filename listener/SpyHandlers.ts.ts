import { DynamoDBSpyEvent } from '../common/spyEvents/DynamoDBSpyEvent';
import { EventBridgeRuleSpyEvent } from '../common/spyEvents/EventBridgeRuleSpyEvent';
import { EventBridgeSpyEvent } from '../common/spyEvents/EventBridgeSpyEvent';
import { FunctionConsole } from '../common/spyEvents/FunctionConsole';
import { FunctionConsoleSpyEvent } from '../common/spyEvents/FunctionConsoleSpyEvent';
import { FunctionContext } from '../common/spyEvents/FunctionContext';
import { FunctionErrorSpyEvent } from '../common/spyEvents/FunctionErrorSpyEvent';
import { FunctionRequestSpyEvent } from '../common/spyEvents/FunctionRequestSpyEvent';
import { FunctionResponseSpyEvent } from '../common/spyEvents/FunctionResponseSpyEvent';
import { S3SpyEvent } from '../common/spyEvents/S3SpyEvent';
import { SnsSubscriptionSpyEvent } from '../common/spyEvents/SnsSubscriptionSpyEvent';
import { SnsTopicSpyEvent } from '../common/spyEvents/SnsTopicSpyEvent';
import { SqsSpyEvent } from '../common/spyEvents/SqsSpyEvent';
import { PrettifyForDisplay } from './PrettifyForDisplay';
import { RecursivePartial } from './RecursivePartial';

type MyJestMatchers = ReturnType<typeof expect>;
type MyJestMatchersWitOnlyFunctions = Pick<
  MyJestMatchers,
  {
    [P in keyof MyJestMatchers]: MyJestMatchers[P] extends (...args: any) => any
      ? P
      : never;
  }[keyof MyJestMatchers]
>;
type JestExpectWithSpyMethods<TSpyEvent, TSpyHandler> = Omit<
  {
    // toMatchObject: (expected: RecursivePartial<TSpyEvent>) => TSpyHandler;
    [K in keyof MyJestMatchersWitOnlyFunctions]: (
      ...args: Parameters<MyJestMatchersWitOnlyFunctions[K]>
    ) => TSpyHandler;
  },
  'toContainEqual' | 'toEqual' | 'toMatchObject' | 'toStrictEqual'
> & {
  /**
   * Used when you want to check that an item is in a list.
   * For testing the items in the list, this matcher recursively checks the
   * equality of all fields, rather than checking for object identity.
   *
   * Optionally, you can provide a type for the expected value via a generic.
   * This is particularly useful for ensuring expected objects have the right structure.
   */

  toContainEqual(expected: TSpyEvent): TSpyHandler;
  /**
   * Used when you want to check that two objects have the same value.
   * This matcher recursively checks the equality of all fields, rather than checking for object identity.
   *
   * Optionally, you can provide a type for the expected value via a generic.
   * This is particularly useful for ensuring expected objects have the right structure.
   */

  toEqual(expected: TSpyEvent): TSpyHandler;

  /**
   * Used to check that a JavaScript object matches a subset of the properties of an object
   *
   * Optionally, you can provide an object to use as Generic type for the expected value.
   * This ensures that the matching object matches the structure of the provided object-like type.
   *
   * @example
   *
   * type House = {
   *   bath: boolean;
   *   bedrooms: number;
   *   kitchen: {
   *     amenities: string[];
   *     area: number;
   *     wallColor: string;
   *   }
   * };
   *
   * expect(desiredHouse).toMatchObject<House>({...standardHouse, kitchen: {area: 20}}) // wherein standardHouse is some base object of type House
   */

  toMatchObject(
    expected: PrettifyForDisplay<RecursivePartial<TSpyEvent>>
  ): TSpyHandler;

  /**
   * Use to test that objects have the same types as well as structure.
   *
   * Optionally, you can provide a type for the expected value via a generic.
   * This is particularly useful for ensuring expected objects have the right structure.
   */

  toStrictEqual(expected: TSpyEvent): TSpyHandler;
};

export interface DynamoDBSpyHandler<TData = any>
  extends JestExpectWithSpyMethods<
    DynamoDBSpyEvent<TData>,
    DynamoDBSpyHandler<TData>
  > {
  getData: () => PrettifyForDisplay<DynamoDBSpyEvent<TData>>;
}
export interface EventBridgeRuleSpyHandler<TData = any>
  extends JestExpectWithSpyMethods<
    EventBridgeRuleSpyEvent<TData>,
    EventBridgeRuleSpyHandler<TData>
  > {
  getData: () => PrettifyForDisplay<EventBridgeRuleSpyEvent<TData>>;
}
export interface EventBridgeSpyHandler<TData = any>
  extends JestExpectWithSpyMethods<
    EventBridgeSpyEvent<TData>,
    EventBridgeSpyHandler<TData>
  > {
  getData: () => PrettifyForDisplay<EventBridgeSpyEvent<TData>>;
}

export interface FunctionBaseSpyHandler<TData = any>
  extends JestExpectWithSpyMethods<
    FunctionRequestSpyEvent<TData>,
    FunctionRequestSpyHandler<TData>
  > {
  followedByError: (
    // This implementation confuses TypeScript which does not accurately display the type.
    param?: {
      condition?: (event: {
        spyEventType: 'FunctionError';
        request: TData;
        context: FunctionContext;
        error: any;
      }) => boolean;
      timoutMs?: number;
    }
  ) => Promise<
    FunctionBaseSpyHandler<TData> &
      FunctionErrorSpyHandler<TData> &
      JestExpectWithSpyMethods<
        FunctionErrorSpyEvent<TData>,
        FunctionRequestSpyHandler<TData>
      >
  >;

  followedByConsole: (
    // This implementation confuses TypeScript which does not accurately display the type.
    param?: {
      condition?: (event: {
        spyEventType: 'FunctionConsole';
        request: TData;
        context: FunctionContext;
        console: FunctionConsole;
      }) => boolean;
      timoutMs?: number;
    }
  ) => Promise<
    FunctionBaseSpyHandler<TData> &
      FunctionConsoleSpyHandler<TData> &
      JestExpectWithSpyMethods<
        FunctionConsoleSpyEvent<TData>,
        FunctionRequestSpyHandler<TData>
      >
  >;

  followedByResponse: <TDataResponse = any>(
    // This implementation confuses TypeScript which does not accurately display the type.
    // Leave it for reference!
    // param: PrettifyForDisplay<
    //   WaitForParams<
    //     PrettifyForDisplay<FunctionResponseSpyEvent<TData, TDataResponse>>
    //   >
    // >
    param?: {
      condition?: (event: {
        spyEventType: 'FunctionResponse';
        request: TData;
        context: FunctionContext;
        response: TDataResponse;
      }) => boolean;
      timoutMs?: number;
    }
  ) => Promise<
    FunctionResponseSpyHandler<TData> &
      JestExpectWithSpyMethods<
        FunctionRequestSpyEvent<TData>,
        FunctionRequestSpyHandler<TData>
      >
  >;
}

export interface FunctionRequestSpyHandler<TData = any>
  extends FunctionBaseSpyHandler<TData> {
  getData: () => PrettifyForDisplay<FunctionRequestSpyEvent<TData>>;
}

export interface FunctionConsoleSpyHandler<TData = any>
  extends JestExpectWithSpyMethods<
    FunctionConsoleSpyEvent<TData>,
    FunctionConsoleSpyHandler<TData>
  > {
  getData: () => PrettifyForDisplay<FunctionConsoleSpyEvent<TData>>;

  followedByError: (
    // This implementation confuses TypeScript which does not accurately display the type.
    param?: {
      condition?: (event: {
        spyEventType: 'FunctionError';
        request: TData;
        context: FunctionContext;
        error: any;
      }) => boolean;
      timoutMs?: number;
    }
  ) => Promise<
    FunctionBaseSpyHandler<TData> &
      FunctionErrorSpyHandler<TData> &
      JestExpectWithSpyMethods<
        FunctionErrorSpyEvent<TData>,
        FunctionRequestSpyHandler<TData>
      >
  >;

  followedByConsole: (
    // This implementation confuses TypeScript which does not accurately display the type.
    param?: {
      condition?: (event: {
        spyEventType: 'FunctionConsole';
        request: TData;
        context: FunctionContext;
        console: FunctionConsole;
      }) => boolean;
      timoutMs?: number;
    }
  ) => Promise<
    FunctionBaseSpyHandler<TData> &
      FunctionConsoleSpyHandler<TData> &
      JestExpectWithSpyMethods<
        FunctionConsoleSpyEvent<TData>,
        FunctionRequestSpyHandler<TData>
      >
  >;

  followedByResponse: <TDataResponse = any>(
    // This implementation confuses TypeScript which does not accurately display the type.
    // Leave it for reference!
    // param: PrettifyForDisplay<
    //   WaitForParams<
    //     PrettifyForDisplay<FunctionResponseSpyEvent<TData, TDataResponse>>
    //   >
    // >
    param?: {
      condition?: (event: {
        spyEventType: 'FunctionResponse';
        request: TData;
        context: FunctionContext;
        response: TDataResponse;
      }) => boolean;
      timoutMs?: number;
    }
  ) => Promise<
    FunctionResponseSpyHandler<TData> &
      JestExpectWithSpyMethods<
        FunctionRequestSpyEvent<TData>,
        FunctionRequestSpyHandler<TData>
      >
  >;
}

export interface FunctionErrorSpyHandler<TData = any>
  extends JestExpectWithSpyMethods<
    FunctionErrorSpyEvent<TData>,
    FunctionErrorSpyHandler<TData>
  > {
  getData: () => PrettifyForDisplay<FunctionErrorSpyEvent<TData>>;
}

export interface FunctionResponseSpyHandler<TData = any>
  extends JestExpectWithSpyMethods<
    FunctionResponseSpyEvent<TData>,
    FunctionResponseSpyHandler<TData>
  > {
  getData: () => PrettifyForDisplay<FunctionResponseSpyEvent<TData>>;
}

export interface S3SpyHandler
  extends JestExpectWithSpyMethods<S3SpyEvent, S3SpyHandler> {
  getData: () => S3SpyEvent;
}
export interface SnsSubscriptionSpyHandler<TData = any>
  extends JestExpectWithSpyMethods<
    SnsSubscriptionSpyEvent<TData>,
    SnsSubscriptionSpyHandler<TData>
  > {
  getData: () => PrettifyForDisplay<SnsSubscriptionSpyEvent<TData>>;
}
export interface SnsTopicSpyHandler<TData = any>
  extends JestExpectWithSpyMethods<
    SnsTopicSpyEvent<TData>,
    SnsTopicSpyHandler<TData>
  > {
  getData: () => PrettifyForDisplay<SnsTopicSpyEvent<TData>>;
}
export interface SqsSpyHandler<TData = any>
  extends JestExpectWithSpyMethods<SqsSpyEvent<TData>, SqsSpyHandler<TData>> {
  getData: () => PrettifyForDisplay<SqsSpyEvent<TData>>;
}
