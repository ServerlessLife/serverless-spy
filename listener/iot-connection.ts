import { IoTClient, DescribeEndpointCommand } from '@aws-sdk/client-iot';
import { device } from 'aws-iot-device-sdk';

export const SSPY_TOPIC = 'sspy';

export type fragment = { id: string; index: number; count: number; data: any };

function createLog(debugMode: boolean) {
  return (message: string, ...optionalParams: any[]) => {
    if (debugMode) {
      console.debug('SSPY', message, ...optionalParams);
    }
  };
}

function createErrorLog() {
  return (message: string, ...optionalParams: any[]) => {
    console.error('SSPY', message, ...optionalParams);
  };
}

export async function getConnection(debugMode: boolean): Promise<device> {
  const log = createLog(debugMode);
  const logError = createErrorLog();
  log('Getting IoT endpoint');
  let response: any;
  try {
    const iotClient = new IoTClient({});
    response = await iotClient.send(
      new DescribeEndpointCommand({
        endpointType: 'iot:Data-ATS',
      })
    );
  } catch (e) {
    logError('failed to get endpoint', e);
    throw e;
  }
  log('Using IoT endpoint:', response.endpointAddress);

  if (!response.endpointAddress) {
    logError('No IoT endpoint could be found');
    throw new Error('IoT Endpoint address not found');
  }

  const connection = new device({
    protocol: 'wss',
    host: response.endpointAddress,
    region: process.env['AWS_REGION'],
    reconnectPeriod: 1,
  });

  connection.on('connect', () => {
    log('IoT connected');
  });

  connection.on('error', (err) => {
    logError('IoT error', err);
  });

  connection.on('close', () => {
    log('IoT closed');
  });

  connection.on('reconnect', () => {
    log('IoT reconnected');
  });

  return connection;
}
