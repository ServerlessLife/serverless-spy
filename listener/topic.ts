export const SSPY_TOPIC = 'sspy';

export function getTopic(scope: string) {
  return `${SSPY_TOPIC}/${scope}`;
}
