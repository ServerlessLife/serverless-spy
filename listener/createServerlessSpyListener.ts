import { ServerlessSpyListenerParams } from './ServerlessSpyListenerParams';
import { WsListener } from './WsListener';

export async function createServerlessSpyListener<TSpyEvents>(
  params: ServerlessSpyListenerParams
) {
  const wsListener = new WsListener<TSpyEvents>();
  let resolve, reject: ((value: void | PromiseLike<void>) => void) | undefined;
  const promise = new Promise<void>((res, rej) => {
    resolve = res;
    reject = rej;
  });
  await wsListener.start({
    ...params,
    connectionOpenResolve: params.connectionOpenResolve || resolve,
    connectionOpenReject: params.connectionOpenReject || reject,
  });

  const proxy = wsListener.createProxy();
  await promise;
  return proxy;
}
