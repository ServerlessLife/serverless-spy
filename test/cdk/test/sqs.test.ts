import * as fs from 'fs';
import * as path from 'path';
import { SendMessageCommand, SQSClient } from '@aws-sdk/client-sqs';
import { v4 as uuidv4 } from 'uuid';
import { TestData } from './TestData';
import { createServerlessSpyListener } from '../../../listener/createServerlessSpyListener';
import { ServerlessSpyListener } from '../../../listener/ServerlessSpyListener';
import { ServerlessSpyEvents } from '../serverlessSpyEvents/ServerlessSpyEventsSqs';

jest.setTimeout(30000);

describe('SQS with option spySqsWithNoSubscriptionAndDropAllMessages', () => {
  const exportLocation = path.join(__dirname, '../cdkOutput.json');
  let serverlessSpyListener: ServerlessSpyListener<ServerlessSpyEvents>;

  if (!fs.existsSync(exportLocation)) {
    throw new Error(`File ${exportLocation} does not exists.`);
  }
  const output = JSON.parse(fs.readFileSync(exportLocation).toString())[
    'ServerlessSpySqs'
  ];

  beforeEach(async () => {
    serverlessSpyListener =
      await createServerlessSpyListener<ServerlessSpyEvents>({
        scope: 'ServerlessSpySqs',
        serverlessSpyWsUrl: output.ServerlessSpyWsUrl,
      });
  });

  afterEach(async () => {
    serverlessSpyListener.stop();
  });

  test('Basic test', async () => {
    const id = uuidv4();
    const data = <TestData>{
      id,
      message: 'Hello',
    };

    const sqsClient = new SQSClient({});
    const sendMessageCommand = new SendMessageCommand({
      MessageBody: JSON.stringify(data),
      QueueUrl: output.QueueUrlMyQueue,
    });
    await sqsClient.send(sendMessageCommand);

    (
      await serverlessSpyListener.waitForSqsMyQueue<TestData>({
        condition: (d) => d.body.id === id,
      })
    ).toMatchObject({ body: data });
  });
});
