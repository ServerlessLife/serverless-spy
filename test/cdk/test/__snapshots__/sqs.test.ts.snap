// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SQS with option spySqsWithNoSubscriptionAndDropAllMessages Snapshot 1`] = `
{
  "Outputs": {
    "QueueUrlMyQueue": {
      "Value": {
        "Ref": "MyQueueE6CA6235",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "MyQueueE6CA6235": {
      "DeletionPolicy": "Delete",
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "ServerlessSpyExtensionEA4873A0": {
      "Properties": {
        "CompatibleArchitectures": [
          "x86_64",
        ],
        "CompatibleRuntimes": [
          "nodejs18.x",
        ],
        "Content": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "babbb172503f6a6ffc178f8d107bae18937392efdbc57c9a729dd35ef6228e53.zip",
        },
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "ServerlessSpyMyQueueSqsSubscriptionAndDropAllMessages3748840C": {
      "DependsOn": [
        "ServerlessSpyMyQueueSqsSubscriptionAndDropAllMessagesServiceRoleDefaultPolicy986A7CEA",
        "ServerlessSpyMyQueueSqsSubscriptionAndDropAllMessagesServiceRoleE8496CCC",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "017c4a8f4be7b88fdbd3f2870a840157cf42e67cb2c44d8b9e81f3995d6ea901.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_LAMBDA_EXEC_WRAPPER": "/opt/spy-wrapper",
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "NODE_OPTIONS": "--enable-source-maps",
            "SSPY_DEBUG": "true",
            "SSPY_INFRA_MAPPING": {
              "Fn::Join": [
                "",
                [
                  "{"",
                  {
                    "Fn::GetAtt": [
                      "MyQueueE6CA6235",
                      "Arn",
                    ],
                  },
                  "":"Sqs#MyQueue"}",
                ],
              ],
            },
            "SSPY_ROOT_STACK": "Test",
            "SSPY_SUBSCRIBED_TO_SQS": "true",
          },
        },
        "Handler": "index.handler",
        "Layers": [
          {
            "Ref": "ServerlessSpyExtensionEA4873A0",
          },
          {
            "Ref": "ServerlessSpySpyWrapperLayerBE5627BF",
          },
        ],
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "ServerlessSpyMyQueueSqsSubscriptionAndDropAllMessagesServiceRoleE8496CCC",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 5,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ServerlessSpyMyQueueSqsSubscriptionAndDropAllMessagesServiceRoleDefaultPolicy986A7CEA": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "MyQueueE6CA6235",
                  "Arn",
                ],
              },
            },
            {
              "Action": "iot:*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ServerlessSpyMyQueueSqsSubscriptionAndDropAllMessagesServiceRoleDefaultPolicy986A7CEA",
        "Roles": [
          {
            "Ref": "ServerlessSpyMyQueueSqsSubscriptionAndDropAllMessagesServiceRoleE8496CCC",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ServerlessSpyMyQueueSqsSubscriptionAndDropAllMessagesServiceRoleE8496CCC": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ServerlessSpyMyQueueSqsSubscriptionAndDropAllMessagesSqsEventSourceTestMyQueueE040091D6114B6E3": {
      "Properties": {
        "EventSourceArn": {
          "Fn::GetAtt": [
            "MyQueueE6CA6235",
            "Arn",
          ],
        },
        "FunctionName": {
          "Ref": "ServerlessSpyMyQueueSqsSubscriptionAndDropAllMessages3748840C",
        },
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "ServerlessSpySpyWrapperLayerBE5627BF": {
      "Properties": {
        "Content": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "bf2df0858ecef77ac31b1994b46cd25f69a84fea5d58e9b5e5c6e458c33012f6.zip",
        },
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "ServerlessSpySubscription0036DF4AB": {
      "DependsOn": [
        "ServerlessSpySubscription0ServiceRoleDefaultPolicyD8FA77F4",
        "ServerlessSpySubscription0ServiceRoleF955D76C",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "2fc02cc64fedb8a5c7c3049c58656172132663d36b541073ada7da4078957efa.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "NODE_OPTIONS": "--enable-source-maps",
            "SSPY_INFRA_MAPPING": "{}",
            "SSPY_ROOT_STACK": "Test",
          },
        },
        "Handler": "index.handler",
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "ServerlessSpySubscription0ServiceRoleF955D76C",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 5,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ServerlessSpySubscription0ServiceRoleDefaultPolicyD8FA77F4": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "iot:*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ServerlessSpySubscription0ServiceRoleDefaultPolicyD8FA77F4",
        "Roles": [
          {
            "Ref": "ServerlessSpySubscription0ServiceRoleF955D76C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ServerlessSpySubscription0ServiceRoleF955D76C": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
